components/layout/query.rs

```diff
@@ -37,16 +40,111 @@ use style::values::specified::GenericGridTemplateComponent;
 use style::values::specified::box_::DisplayInside;
 use style::values::specified::text::TextTransformCase;
 use style_traits::{ParsingMode, ToCss};
+use webrender_api::units::LayoutTransform;
 
 use crate::ArcRefCell;
 use crate::dom::NodeExt;
 use crate::flow::inline::construct::{TextTransformation, WhitespaceCollapse, capitalize_string};
 use crate::fragment_tree::{
     BoxFragment, Fragment, FragmentFlags, FragmentTree, SpecificLayoutInfo,
 };
+use crate::style_ext::TransformExt;
 use crate::taffy::SpecificTaffyGridInfo;
 
-pub fn process_content_box_request(node: ServoLayoutNode<'_>) -> Option<Rect<Au>> {
+/// Specific helper for post composite query, query that requires the consideration of transform,
+/// mapping of coordinates, scroll offset, etc. It provide some encapsulation of borrowed structures
+/// required in these kind of queries and some useful utilities of it.
+pub(crate) struct PostCompositeQueryHelper<'dom> {
+    scroll_tree: Option<&'dom ScrollTree>,
+}
+
+impl<'dom> PostCompositeQueryHelper<'dom> {
+    pub fn new(scroll_tree: Option<&'dom ScrollTree>) -> Self {
+        PostCompositeQueryHelper { scroll_tree }
+    }
+
+    fn retrieve_box_fragment(fragment: &Fragment) -> Option<&ArcRefCell<BoxFragment>> {
+        match fragment {
+            Fragment::Box(box_fragment) | Fragment::Float(box_fragment) => Some(box_fragment),
+            _ => None,
+        }
+    }
+
+    fn get_scroll_node_id_for_query(&self, node: ServoLayoutNode<'_>) -> Option<ScrollTreeNodeId> {
+        let fragments = node.fragments_for_pseudo(None);
+        let box_fragment = fragments
+            .first()
+            .and_then(PostCompositeQueryHelper::retrieve_box_fragment)?
+            .borrow();
+        *box_fragment.scroll_tree_node_id_for_query.borrow()
+    }
+
+    /// Get a scroll node that would represents this [ServoLayoutNode]'s transform and
+    /// calculate its transfrom from its root scroll node to the scroll node.
+    fn layout_node_to_root_transform(&self, node: ServoLayoutNode<'_>) -> Option<LayoutTransform> {
+        let scroll_tree_node_id = self.get_scroll_node_id_for_query(node)?;
+
+        Some(self.scroll_node_to_root_transform(&scroll_tree_node_id))
+    }
+
+    /// Traverse a scroll node to its root to calculate the transform.
+    // TODO(stevennovaryo): Add caching mechanism for this.
+    fn scroll_node_to_root_transform(&self, node_id: &ScrollTreeNodeId) -> LayoutTransform {
```

Comment:
Let's just make this a free function called `fn cumulative_scroll_node_transform_for_node(node: ServoLayoutNode<'_>)` and combine it with `get_scroll_node_id_for_query`. The name `PostCompositeQueryHelper` is really general and this is really just doing one thing (getting the cumulative scroll node transformation for a node).

Reply:
Sorry, the scope and state of these functions does seems quite inappropriate. I have changed `cumulative_scroll_node_transform_for_node` to be a free function of `PostCompositeQueryHelper`. That feels appropriate since it is still a subset of post composite queries. I have also combined `get_scroll_node_id_for_query` with `layout_node_to_root_transform`, seems that is the most appropriate approach in the current state.